File: solana/veritas-curation/programs/veritas-curation/src/pool_factory/instructions/initialize_factory.rs
use anchor_lang::prelude::*;
use anchor_lang::system_program;

use crate::pool_factory::{
    state::{
        PoolFactory,
        FACTORY_SEED,
        DEFAULT_F,
        DEFAULT_BETA_NUM,
        DEFAULT_BETA_DEN,
        DEFAULT_P0,
        DEFAULT_MIN_INITIAL_DEPOSIT,
        DEFAULT_MIN_SETTLE_INTERVAL,
    },
    events::FactoryInitializedEvent,
    errors::FactoryError,
};
use crate::program::VeritasCuration;

/// Initialize the singleton factory PDA with protocol authority and fee configuration
pub fn initialize_factory(
    ctx: Context<InitializeFactory>,
    protocol_authority: Pubkey,
    custodian: Pubkey,
    total_fee_bps: u16,
    creator_split_bps: u16,
    protocol_treasury: Pubkey,
) -> Result<()> {
    // Validate upgrade authority
    let program_data_bytes = ctx.accounts.program_data.try_borrow_data()?;
    if program_data_bytes.len() < 45 {
        return Err(FactoryError::InvalidProgramData.into());
    }

    // Deserialize: first 4 bytes = discriminator, next 8 = slot, next 1 = Option tag, next 32 = Pubkey
    let upgrade_authority_option = if program_data_bytes[12] == 0 {
        None
    } else {
        let mut pubkey_bytes = [0u8; 32];
        pubkey_bytes.copy_from_slice(&program_data_bytes[13..45]);
        Some(Pubkey::new_from_array(pubkey_bytes))
    };

    require!(
        upgrade_authority_option == Some(ctx.accounts.upgrade_authority.key()),
        FactoryError::InvalidUpgradeAuthority
    );

    let factory = &mut ctx.accounts.factory;
    let clock = Clock::get()?;

    // Validate authorities
    require!(
        protocol_authority != Pubkey::default(),
        FactoryError::InvalidAuthority
    );
    require!(
        custodian != Pubkey::default(),
        FactoryError::InvalidAuthority
    );
    require!(
        protocol_treasury != Pubkey::default(),
        FactoryError::InvalidAuthority
    );
    require!(
        protocol_authority != system_program::ID,
        FactoryError::InvalidAuthority
    );
    require!(
        protocol_treasury != system_program::ID,
        FactoryError::InvalidAuthority
    );

    // Validate fee configuration
    require!(
        creator_split_bps <= 10000,
        FactoryError::InvalidCreatorSplit
    );

    // Initialize state
    factory.protocol_authority = protocol_authority;
    factory.total_pools = 0;
    factory.total_fee_bps = total_fee_bps;
    factory.creator_split_bps = creator_split_bps;
    factory.protocol_treasury = protocol_treasury;
    factory._padding_fee = [0; 2];
    factory.default_f = DEFAULT_F;
    factory.default_beta_num = DEFAULT_BETA_NUM;
    factory.default_beta_den = DEFAULT_BETA_DEN;
    factory.default_p0 = DEFAULT_P0;
    factory.min_initial_deposit = DEFAULT_MIN_INITIAL_DEPOSIT;
    factory.min_settle_interval = DEFAULT_MIN_SETTLE_INTERVAL;
    factory.custodian = custodian;
    factory.bump = ctx.bumps.factory;

    emit!(FactoryInitializedEvent {
        factory: factory.key(),
        protocol_authority,
        custodian,
        total_fee_bps,
        creator_split_bps,
        protocol_treasury,
        timestamp: clock.unix_timestamp,
    });

    Ok(())
}

#[derive(Accounts)]
pub struct InitializeFactory<'info> {
    #[account(
        init,
        payer = payer,
        space = 8 + PoolFactory::LEN,
        seeds = [FACTORY_SEED],
        bump
    )]
    pub factory: Account<'info, PoolFactory>,

    pub upgrade_authority: Signer<'info>,

    #[account(constraint = program.programdata_address()? == Some(program_data.key()))]
    pub program: Program<'info, VeritasCuration>,

    /// CHECK: Program data account validated in handler
    pub program_data: AccountInfo<'info>,

    #[account(mut)]
    pub payer: Signer<'info>,

    pub system_program: Program<'info, System>,
}

File: solana/veritas-curation/scripts/initialize-factory.ts
/**
 * Initialize PoolFactory
 *
 * This script:
 * 1. Initializes the PoolFactory singleton
 * 2. Sets factory authority (administrative control)
 * 3. Sets pool authority (operational control - signs trades, settlements)
 * 4. Sets custodian reference
 * 5. Sets default ICBS parameters (F=3, β=0.5)
 *
 * Usage: npx ts-node scripts/initialize-factory.ts
 */

import * as anchor from "@coral-xyz/anchor";
import { Program } from "@coral-xyz/anchor";
import { PublicKey, SystemProgram } from "@solana/web3.js";
import { VeritasCuration } from "../target/types/veritas_curation";
import fs from "fs";
import path from "path";
import { loadProtocolAuthority } from "./load-authority";

async function main() {
  console.log("🚀 Initializing PoolFactory...\n");

  // Setup provider
  const provider = anchor.AnchorProvider.env();
  anchor.setProvider(provider);

  const program = anchor.workspace.VeritasCuration as Program<VeritasCuration>;
  const wallet = provider.wallet;

  console.log("📋 Program ID:", program.programId.toString());
  console.log("👤 Wallet:", wallet.publicKey.toString());
  console.log("");

  // Derive factory PDA
  const [factoryPda, factoryBump] = PublicKey.findProgramAddressSync(
    [Buffer.from("factory")],
    program.programId
  );
  console.log("🏭 Factory PDA:", factoryPda.toString());

  // Derive custodian PDA (must exist)
  const [custodianPda] = PublicKey.findProgramAddressSync(
    [Buffer.from("custodian")],
    program.programId
  );
  console.log("🏛️  Custodian PDA:", custodianPda.toString());
  console.log("");

  // Verify custodian exists
  try {
    await program.account.veritasCustodian.fetch(custodianPda);
    console.log("✅ Custodian verified");
  } catch (err) {
    console.error("❌ Custodian not found! Run initialize-custodian.ts first.");
    process.exit(1);
  }

  // Check if already initialized
  try {
    const factoryAccount = await program.account.poolFactory.fetch(factoryPda);
    console.log("⚠️  Factory already initialized!");
    console.log("   Protocol Authority:", factoryAccount.protocolAuthority.toString());
    console.log("   Protocol Treasury:", factoryAccount.protocolTreasury.toString());
    console.log("   Custodian:", factoryAccount.custodian.toString());
    console.log("   Total Pools:", factoryAccount.totalPools.toString());
    console.log("   Total Fee:", `${factoryAccount.totalFeeBps} bps`);
    console.log("   Creator Split:", `${factoryAccount.creatorSplitBps} bps`);
    console.log("   Default F:", factoryAccount.defaultF);
    console.log("   Default β:", `${factoryAccount.defaultBetaNum}/${factoryAccount.defaultBetaDen}`);
    console.log("");
    console.log("✅ Factory is ready!");
    return;
  } catch (err) {
    // Not initialized yet, continue
    console.log("📝 Factory not yet initialized, proceeding...");
  }

  // Initialize factory
  console.log("⚙️  Sending initialize transaction...");

  // Load protocol authority from environment variable
  const protocolAuthorityKeypair = loadProtocolAuthority();
  const protocolAuthority = protocolAuthorityKeypair.publicKey;

  // Use upgrade authority as protocol treasury for mainnet
  const protocolTreasury = wallet.publicKey;

  // Fee configuration from deployment.config.json:
  // 0.5% total fee (50 bps), 100% to creator (10000 bps of 10000)
  const totalFeeBps = 50;
  const creatorSplitBps = 10000;

  // Get program data PDA
  const [programDataAddress] = PublicKey.findProgramAddressSync(
    [program.programId.toBuffer()],
    new PublicKey("BPFLoaderUpgradeab1e11111111111111111111111")
  );

  try {
    const tx = await program.methods
      .initializeFactory(
        protocolAuthority,
        custodianPda,
        totalFeeBps,
        creatorSplitBps,
        protocolTreasury
      )
      .accounts({
        upgradeAuthority: wallet.publicKey,
        programData: programDataAddress,
        payer: wallet.publicKey,
      })
      .rpc();

    console.log("✅ Transaction successful!");
    console.log("   Signature:", tx);
    console.log("");

    // Verify initialization
    const factoryAccount = await program.account.poolFactory.fetch(factoryPda);
    console.log("✅ Factory initialized successfully!");
    console.log("");
    console.log("📊 Factory Details:");
    console.log("   Address:", factoryPda.toString());
    console.log("   Protocol Authority:", factoryAccount.protocolAuthority.toString());
    console.log("   Protocol Treasury:", factoryAccount.protocolTreasury.toString());
    console.log("   Custodian:", factoryAccount.custodian.toString());
    console.log("   Total Pools:", factoryAccount.totalPools.toString());
    console.log("   Total Fee:", `${factoryAccount.totalFeeBps} bps`);
    console.log("   Creator Split:", `${factoryAccount.creatorSplitBps} bps`);
    console.log("");
    console.log("📐 Default ICBS Parameters:");
    console.log("   F (growth exponent):", factoryAccount.defaultF);
    console.log("   β (coupling coefficient):", `${factoryAccount.defaultBetaNum}/${factoryAccount.defaultBetaDen} = ${factoryAccount.defaultBetaNum / factoryAccount.defaultBetaDen}`);
    console.log("   Min Initial Deposit:", factoryAccount.minInitialDeposit.toString(), "lamports");
    console.log("   Min Settle Interval:", factoryAccount.minSettleInterval.toString(), "seconds");
    console.log("");

    // Save addresses to file for later use
    const outputPath = path.join(__dirname, "../.local-deployment.json");
    let deploymentData: any = {};

    if (fs.existsSync(outputPath)) {
      deploymentData = JSON.parse(fs.readFileSync(outputPath, "utf-8"));
    }

    deploymentData.factory = {
      address: factoryPda.toString(),
      protocolAuthority: factoryAccount.protocolAuthority.toString(),
      protocolTreasury: factoryAccount.protocolTreasury.toString(),
      custodian: factoryAccount.custodian.toString(),
      totalFeeBps: factoryAccount.totalFeeBps,
      creatorSplitBps: factoryAccount.creatorSplitBps,
      defaultF: factoryAccount.defaultF,
      defaultBetaNum: factoryAccount.defaultBetaNum,
      defaultBetaDen: factoryAccount.defaultBetaDen,
    };

    fs.writeFileSync(outputPath, JSON.stringify(deploymentData, null, 2));
    console.log("💾 Saved factory addresses to:", outputPath);
    console.log("");

    console.log("🎉 Setup complete!");
    console.log("");
    console.log("📝 Next steps:");
    console.log("   1. Use create-test-pool.ts to create a test pool");
    console.log("   2. Deploy market on the pool");
    console.log("   3. Execute test trades");

  } catch (error) {
    console.error("❌ Error initializing factory:", error);
    throw error;
  }
}

main()
  .then(() => process.exit(0))
  .catch((error) => {
    console.error(error);
    process.exit(1);
  });
